cpp/src/hdbscan/detail/utils.h

```diff
@@ -0,0 +1,139 @@
+/*
+ * Copyright (c) 2021, NVIDIA CORPORATION.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+#pragma once
+
+#include <label/classlabels.cuh>
+
+#include <cub/cub.cuh>
+
+#include <raft/cudart_utils.h>
+
+#include <raft/sparse/op/sort.h>
+#include <raft/sparse/convert/csr.cuh>
+
+#include <cuml/cluster/hdbscan.hpp>
+
+#include <raft/label/classlabels.cuh>
+
+#include <algorithm>
+
+#include <hdbscan/condensed_hierarchy.cu>
+
+#include <thrust/execution_policy.h>
+#include <thrust/for_each.h>
+#include <thrust/reduce.h>
+#include <thrust/sort.h>
+#include <thrust/transform.h>
+
+#include <rmm/device_uvector.hpp>
+#include <rmm/exec_policy.hpp>
+
+namespace ML {
+namespace HDBSCAN {
+namespace detail {
+namespace Utils {
+
+/**
+ * Invokes a cub segmented reduce function over a CSR data array
+ * using the indptr as segment offsets
+ * @tparam value_idx
+ * @tparam value_t
+ * @tparam CUBReduceFunc
+ * @param[in] in data array (size offsets[n_segments]+1)
+ * @param[out] out output data array (size offsets[n_segmented]+1)
+ * @param[in] n_segments number of segments in offsets array
+ * @param[in] offsets array of segment offsets (size n_segments+1)
+ * @param[in] stream cuda stream for ordering operations
+ * @param[in] cub_reduce_func segmented reduction function
+ */
+template <typename value_idx, typename value_t, typename CUBReduceFunc>
+void cub_segmented_reduce(const value_t *in, value_t *out, int n_segments,
+                          const value_idx *offsets, cudaStream_t stream,
+                          CUBReduceFunc cub_reduce_func) {
+  void *d_temp_storage = NULL;
+  size_t temp_storage_bytes = 0;
+  cub_reduce_func(d_temp_storage, temp_storage_bytes, in, out, n_segments,
+                  offsets, offsets + 1, stream, false);
+  CUDA_CHECK(cudaMalloc(&d_temp_storage, temp_storage_bytes));
```

Comment:
How to use it? I foresee two problems:

1. The cub APIs use a `void *` temporary storage. So, I cannot assign a type as I don't know internally what type/alignment they are expected.
2. From above, it also becomes not possible to gauge the length of the array since the size is in bytes

Reply:
Oh I found an example in the codebase on how to use this. I used `rmm::device_uvector<char>` instead